docs/dev/clusteroperator.md

```diff
@@ -76,7 +76,22 @@ Here are the guarantees components can get when they follow the rules we define:
 There are a set of guarantees components are expected to honor in return:
 
 1. A component doesn't report the `Available` status condition the first time until they are completely rolled out (or within some reasonable percentage if the component must be installed to all nodes)
-2. A component reports `Failing` when it can't accomplish its task. This might be very broad - the API servers are down so I failed to update - or narrow - I couldn't update the last secret. In either case, Failing communicates something is wrong.
+2. A component reports `Degraded` when it can't accomplish its task or if its
+   current state does not match its desired state resulting in a lower quality
+   of service.  For example, if you desire 3 pods, and 2 are running, but 1 is
+   crash-looping, your operator service is possibly still `Available`, but its
+   service is in a `Degraded` state.  A component may be `Degraded` while it is
+   `Progressing` to a new desired state; for example, if only 2 of the 3 desired
+   replicas are achieved.  As a result, it may be normal for a operator during
+   upgrade to temporarily report `Degraded`.  A component may be `Progressing`
```

Comment:
I'm still not 100% on board with this.  I don't want operators flashing between degraded and not degraded.  There needs to be a "within a minute or so, if the condition isn't resolved, an operator should report degraded".  It's acceptable to start by flashing degraded, but I'm going to open bugs to you until during normal upgrades you don't.

I don't think being 2/3 during a rolling upgrade is degraded.  I think being 2/3 because the upgrade can't make progress is degraded.

Reply:
We need to be very careful that people don't interpret Degraded as "normal operation but change is happening".  Degraded is "you either have no idea what is going on" (which is something a development team is expected to fix and bugs will be opened against you) or "something is legitimately failing".

I don't want to enshrine "shrug" as degraded.